<?xml version="1.0"?>
<launch>

    <!-- Name of the UWB sequence -->
    <arg name="bag_file" default="/home/xy/DATASETS/CTGAUPRO_TWR/modified_linktrack_data_infinit1.bag"/>
    <arg name="result_save_path" default="/home/kailai/Documents/results/ctgaupro/c1/"/>
    <arg name="anchor_path" default="/home/xy/DATASETS/CTGAUPRO_TWR/anchor_const1_survey.txt"/>
    

    <!-- Launch the uwb imu fusion node -->
    <node pkg="ctgaupro" type="ctgaupro_gpui" name="ctgaupro_gpui" output="screen">
        
        <param name="auto_exit"  value="1"/>
        <param name="if_save_traj" value="1"/>
        <param name="traj_save_path" value="$(arg result_save_path)"/>

        <!-- Parameters for the Gaussian Process -->
        <param name="gpDt"       value="0.04357"/>
        <param name="gpQr"       value="1.00"/>
        <param name="gpQc"       value="1.00"/>
        
        <!-- UWB anchor position -->
        <param name="anchor_pose_path" value="$(arg anchor_path)"/>
        <param name="linktrack_topic0" value="/nlink_linktrack_nodeframe3_tag0"/>
        <param name="linktrack_topic1" value="/nlink_linktrack_nodeframe3_tag1"/>
        <param name="linktrack_topic2" value="/nlink_linktrack_nodeframe3_tag2"/>
        <param name="linktrack_topic3" value="/nlink_linktrack_nodeframe3_tag3"/>
        <param name="tof_topic"  value="/tof_data" />
        <param name="imu_topic"  value="/imu_data" />
        <param name="gt_topic"   value="/pose_data" /> 
        <param name="fuse_linktrack"  value="1"/>
        <param name="fuse_tof"   value="0"/>
        <param name="fuse_imu"   value="0"/>

        <!-- Parameters for the solver  -->
        <param name="SLIDE_SIZE"  value="2"/> <!-- How many knots to slide -->
        <param name="WINDOW_SIZE" value="20"/> <!-- How many knot length does the sliding window -->
        <param name="w_linktrack" value="100.0"/> <!-- Coefficients for TDOA residuals -->
        <param name="GYR_N" value="2000.0"/> <!-- Coefficients for IMU residuals -->
        <param name="GYR_W" value="100.0"/> <!-- Coefficients for IMU residuals -->
        <param name="ACC_N" value="100.0"/> <!-- Coefficients for IMU residuals -->
        <param name="ACC_W" value="100.0"/> <!-- Coefficients for IMU residuals -->
        <param name="linktrack_loss_thres" value="30.0"/> <!-- Loss function for IMU residuals -->
        <param name="mp_loss_thres" value="100.0"/> <!-- Loss function for motion prior residuals -->

    </node>

   <!-- Visualize -->
    <node pkg="rviz" type="rviz" name="rviz_ui" output="log" required="true"
        args="-d $(find ctgaupro)/launch/ctgaupro_ui.rviz" />        

    <!-- Play the bag file -->
    <node required="true" pkg="rosbag" type="play" name="rosbag_player"
          args="--clock $(arg bag_file) -r 0.5 -s 0.0"
          launch-prefix="bash -c 'sleep 1.0; $0 $@' "/>       

</launch>